group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2904,6 +2908,135 @@ private void validateGenericGroupHeartbeat(
+    public CoordinatorResult<LeaveGroupResponseData, Record> genericGroupLeave(
+        RequestContext context,
+        LeaveGroupRequestData request
+    ) throws UnknownMemberIdException, GroupIdNotFoundException {
+        GenericGroup group = getOrMaybeCreateGenericGroup(request.groupId(), false);
+        if (group.isInState(DEAD)) {
+            return new CoordinatorResult<>(Collections.emptyList(),
+                new LeaveGroupResponseData()
+                    .setErrorCode(COORDINATOR_NOT_AVAILABLE.code())
+            );
+        }
+
+        CoordinatorResult<Void, Record> coordinatorResult = EMPTY_RESULT;
+        List<MemberResponse> memberResponses = new ArrayList<>();
+
+            for (MemberIdentity member : request.members()) {
+                // The LeaveGroup API allows administrative removal of members by GroupInstanceId
+                // in which case we expect the MemberId to be undefined.
+                if (member.memberId().equals(UNKNOWN_MEMBER_ID)) {
+                    if (member.groupInstanceId() != null && group.staticMemberId(member.groupInstanceId()) != null) {
+                        coordinatorResult = removeCurrentMemberFromGenericGroup(
+                            group,
+                            group.staticMemberId(member.groupInstanceId()),
+                            member.reason()
+                        );
+                        memberResponses.add(
+                            new MemberResponse()
+                                .setMemberId(member.memberId())
+                                .setGroupInstanceId(member.groupInstanceId())
+                        );
+                    } else {
+                        memberResponses.add(
+                            new MemberResponse()
+                                .setMemberId(member.memberId())
+                                .setGroupInstanceId(member.groupInstanceId())
+                                .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code())
+                        );
+                    }
+                } else if (group.isPendingMember(member.memberId())) {
+                    coordinatorResult = removePendingMemberAndUpdateGenericGroup(group, member.memberId());
+                    timer.cancel(genericGroupHeartbeatKey(group.groupId(), member.memberId()));
+                    log.info("Pending member {} has left group {} through explicit `LeaveGroup` request.",
+                        member.memberId(), group.groupId());
+
+                    memberResponses.add(
+                        new MemberResponse()
+                            .setMemberId(member.memberId())
+                            .setGroupInstanceId(member.groupInstanceId())
+                    );
+                } else {
+                    try {
+                        group.validateMember(member.memberId(), member.groupInstanceId(), "leave-group");
+                        coordinatorResult = removeCurrentMemberFromGenericGroup(
```

Comment:
The fact that we override `coordinatorResult` for each member worries me. Can't we have issues with this? For instance, let's say that the the first member is removed from the group and the group become empty. This generate a record. Now, the next member in the request is a pending member. Wouldn't we lose the previous record in this case?

Reply:
Added a test case as suggested below (`testJoinedMemberPendingMemberBatchLeaveGroup` and `testJoinedMemberPendingMemberBatchLeaveGroupWithUnknownMember`). Even though the group becomes empty, as there still is a pending member `hasAllMembersJoined()` returns false and does not generate records. Only after the last pending member leaves the group really becomes empty.

So, the leave group request can only produce records in the last valid member.

I understand that this looks unsafe, should we append the coordinatorResults and merge them instead?

Reply:
Yeah, aggregating them would be safer. Is it easy to do?

Reply:
Aggregating would go against the coordinator runtime framework as we would have multiple responses and record lists. 

Instead, I changed the code to check whether we observe more than 1 non-empty coordinator result as we confirmed this 1 (max) non-empty result is the correct and expected behavior. Let me know your thoughts